lib/puppetserver/utils/refactored_client.rb

```diff
@@ -0,0 +1,108 @@
+require 'openssl'
+require 'net/https'
+
+module Puppetserver
+  module Utils
+    # Utilities for doing HTTPS against the CA that wraps Net::HTTP constructs
+    class RefactoredClient
+
+      HEADERS = {
+        'User-Agent'   => 'PuppetserverCaCli',
+        'Content-Type' => 'application/json',
+        'Accept'       => 'application/json'
+      }
+
+      def initialize(localcacert, crl_usage, hostcrl)
+        @store = make_store(localcacert, crl_usage, hostcrl)
+      end
+
+      # Returns a URI-like wrapper around CA specific urls
+      def make_ca_url(host, port, resource_type = nil, certname = nil)
+        URL.new('https', host, port, 'puppet-ca', 'v1', resource_type, certname)
+      end
+
+      # Takes an instance URL (defined lower in the file), and creates a
+      # connection. The given block is passed our own Connection object.
+      # The Connection object should have HTTP verbs defined on it that take
+      # a body (and optional overrides). Returns whatever the block given returned.
+      def with_connection(url, &block)
+        request = ->(conn) { block.call(Connection.new(conn, url)) }
+
+        Net::HTTP.start(url.host, url.port,
+                        use_ssl: true, cert_store: @store,
+                        &request)
+      end
+
+
+    private
+
+      # Helper class that wraps a Net::HTTP connection, a HttpUtilities::URL
+      # and defines methods named after HTTP verbs that are called on the
+      # saved connection, returning a Result.
+      class Connection
+        def initialize(net_http_connection, url_struct)
+          @conn = net_http_connection
+          @url = url_struct
+        end
+
+        def get(url_override = nil)
+          url = url_override || @url
+
+          request = Net::HTTP::Get.new(url.to_uri, HEADERS)
+          result = @conn.request(request)
+
+          Result.new(result.code, result.body)
+        end
+
+        def put(body, url_override = nil)
+          url = url_override || @url
+
+          request = Net::HTTP::Put.new(url.to_uri, HEADERS)
+          request.body = body
+          result = @conn.request(request)
+
+          Result.new(result.code, result.body)
+        end
+      end
+
+      # Just provide the bits of Net::HTTPResponse we care about
+      Result = Struct.new(:code, :body)
+
+      # Like URI, but not... maybe of suspicious value
+      URL = Struct.new(:protocol, :host, :port,
```

Comment:
Why did we decide to not just use `URI` here originally?

Reply:
Not exactly sure.  @justinstoller ?

Reply:
I originally just wrapped all of the things that were net/http or it tightly integrated with. I waffled on whether that was excessive after I wrote it, see my comment above this line.  (: